pkg/controller/garbagecollector/graph_builder.go

```diff
@@ -0,0 +1,362 @@
+		}
+		p.insertNode(newNode)
+		// the underlying delta_fifo may combine a creation and deletion into one event
+		if shouldOrphanDependents(event, accessor) {
+			glog.V(5).Infof("add %s to the attemptToOrphan", newNode.identity)
+			p.attemptToOrphan.Add(newNode)
+		}
+		if startsWaitingForDependentsDeletion(event, accessor) {
+			glog.V(2).Infof("add %s to the attemptToDelete, because it's waiting for its dependents to be deleted", newNode.identity)
+			p.attemptToDelete.Add(newNode)
+			for dep := range newNode.dependents {
+				p.attemptToDelete.Add(dep)
+			}
+		}
+	case (event.eventType == addEvent || event.eventType == updateEvent) && found:
+		// caveat: if GC observes the creation of the dependents later than the
+		// deletion of the owner, then the orphaning finalizer won't be effective.
+		if shouldOrphanDependents(event, accessor) {
+			glog.V(5).Infof("add %s to the attemptToOrphan", existingNode.identity)
+			p.attemptToOrphan.Add(existingNode)
+		}
+		if beingDeleted(accessor) {
+			existingNode.beingDeleted = true
+		}
+		if startsWaitingForDependentsDeletion(event, accessor) {
+			glog.V(2).Infof("add %s to the attemptToDelete, because it's waiting for its dependents to be deleted", existingNode.identity)
+			// if the existingNode is added as a "virtual" node, its deletingDependents field is not properly set, so always set it here.
+			existingNode.deletingDependents = true
+			p.attemptToDelete.Add(existingNode)
+			for dep := range existingNode.dependents {
+				p.attemptToDelete.Add(dep)
+			}
+		}
+		// add/remove owner refs
+		added, removed, changed := referencesDiffs(existingNode.owners, accessor.GetOwnerReferences())
+		if len(added) == 0 && len(removed) == 0 && len(changed) == 0 {
+			glog.V(5).Infof("The updateEvent %#v doesn't change node references, ignore", event)
+			return
+		}
+		if len(changed) != 0 {
+			glog.V(5).Infof("references %#v changed for object %s", changed, existingNode.identity)
+		}
+		p.addUnblockedOwnersToDirtyQueue(removed, changed)
+		// update the node itself
+		existingNode.owners = accessor.GetOwnerReferences()
+		// Add the node to its new owners' dependent lists.
+		p.addDependentToOwners(existingNode, added)
+		// remove the node from the dependent list of node that are no longer in
+		// the node's owners list.
+		p.removeDependentFromOwners(existingNode, removed)
+	case event.eventType == deleteEvent:
+		if !found {
+			glog.V(5).Infof("%v doesn't exist in the graph, this shouldn't happen", accessor.GetUID())
+			return
+		}
+		p.removeNode(existingNode)
+		existingNode.dependentsLock.RLock()
+		defer existingNode.dependentsLock.RUnlock()
+		if len(existingNode.dependents) > 0 {
+			absentOwnerCache.Add(accessor.GetUID())
+		}
+		for dep := range existingNode.dependents {
+			p.attemptToDelete.Add(dep)
+		}
+		for _, owner := range existingNode.owners {
+			ownerNode, found := p.uidToNode.Read(owner.UID)
+			if !found || !ownerNode.deletingDependents {
+				continue
```

Comment:
I added comment on "p.removeNode()" saying it's going to update the graph. I didn't reorder the code because as you commented we'd better get the graph straight before adding things to the queues.